locust/user/task.py

```diff
@@ -80,6 +128,30 @@ def get_tasks_from_base_classes(bases, class_dict):
     
     return new_tasks
 
+def filter_tasks_by_tags(task_holder, include_tags=None, exclude_tags=None, checked={}):
+    """
+    Function used by Environment to recursively remove any tasks/TaskSets from
+    a TaskSet/User that shouldn't be executed according to the tag options
+    """
+
+    def passes_tags(task):
+        if task in checked:
+            return checked[task]
+
+        passing = True
+        if issubclass(type(task), TaskSetMeta):
+            filter_tasks_by_tags(task, include_tags, exclude_tags, checked)
+            passing = len(task.tasks) > 0
+        else:
+            if include_tags is not None:
+                passing &= 'locust_tag_set' in dir(task) and len(task.locust_tag_set & include_tags) > 0
+            if exclude_tags is not None:
+                passing &= 'locust_tag_set' not in dir(task) or len(task.locust_tag_set & exclude_tags) == 0
+
+        checked[task] = passing
+        return passing
+
+    task_holder.tasks = list(filter(passes_tags, task_holder.tasks))
```

Comment:
It feels a bit weird to have a side effect of calling `filter_tasks_by_tags` recursively within the `passes_tags` function. Maybe it would be better to, instead of using `filter()`, just do something like:

```python
new_tags = []
for task in task_holder.tasks:
    ...
task_holder.tasks = new_tags
```
?